###
#   File name : DifferentialMethylation2.R
#   Author    : Hyunjin Kim
#   Date      : Nov 24, 2018
#   Email     : hk2990@cumc.columbia.edu
#   Purpose   : Perform DM analysis on the normalized methylation data
#
#   * THIS CODE IS DIFFERENT FROM "DifferentialMethylation.R" SINCE THIS IS FOR THE OLD DATA
#     SPECIFICALLY, OLD(450K-PERIODONTITIS)+NEW(EPIC-CONTROL) GENERATED BY "Virtual450kContorls.R"
#
#   Workflow
#               https://www.bioconductor.org/packages/devel/workflows/vignettes/methylationArrayAnalysis/inst/doc/methylationArrayAnalysis.html#obtaining-the-data
#
#               1. There are 3 comparisons (Periodontitis vs Healthy, Gingivitis vs Healthy, and Periodontitis vs Gingivitis),
#                  and "limma" is used to get differentially methylated sites.
#                  There is a batch effect among different sample groups, therefore, the effect is considered
#                  as a confoudning factor and included in the design matrix of the linear model.
#                  For the design matrix, [~0+phenotype+samplegroup] is used.
#                  "IlluminaHumanMethylationEPICanno.ilm10b2.hg19" is used to annotate the CpG sites.
#               2. After getting the differentially methylated sites (~850k), "DMRcate" is used to generate
#                  differentially methylated regions. 1000 is used for the lambda parameter in the dmrcate()
#                  function which means gaps >= lambda between significant CpG sites will be in separate DMRs.
#               3. The DMR.plot() draws a plot of the top differentially methylated reguions associated with
#                  chromosome location and nearby genes. It also represents methylation value of the samples.
#               4. Pathway analysis is performed based on genes asoociated with the differentially
#                  methylated CpG sites. Go and Kegg databases are used.
#
#   Instruction
#               1. Source("DifferentialMethylation2.R")
#               2. Run the function "dma" - specify the input file path (RDA file) and output directory
#               3. Differentially methylated results will be generated under the output directory
#
#   Example
#               > source("The_directory_of_DifferentialMethylation2.R/DifferentialMethylation2.R")
#               > dma2(inputFilelPath="./data/450k/virtual_450k_with_controls/panos_450k_virtual_fullset.rda",
#                      fdrThreshold=0.05,
#                      dmrPrintNum=5,
#                      outputDir="./results/DMA/")
###

dma2 <- function(inputFilelPath="//isilon.c2b2.columbia.edu/ifs/archive/shares/bisr/Papapanou/Sept_2018/PreprocessedData/450k/virtual_450k_with_controls/panos_450k_virtual_fullset.rda",
                 fdrThreshold=0.05,
                 dmrPrintNum=5,
                 outputDir="//isilon.c2b2.columbia.edu/ifs/archive/shares/bisr/Papapanou/Sept_2018/DifferentialMethylation/") {
  
  ### load libraries
  if(!require(limma)) {
    source("https://bioconductor.org/biocLite.R")
    biocLite("limma")
    library(limma)
  }
  if(!require(minfi)) {
    source("https://bioconductor.org/biocLite.R")
    biocLite("minfi")
    library(minfi)
  }
  if(!require(IlluminaHumanMethylation450kanno.ilmn12.hg19)) {
    source("https://bioconductor.org/biocLite.R")
    biocLite("IlluminaHumanMethylation450kanno.ilmn12.hg19")
    library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
  }
  if(!require(DMRcate)) {
    source("https://bioconductor.org/biocLite.R")
    biocLite("DMRcate")
    library(DMRcate)
  }
  if(!require(missMethyl)) {
    source("https://bioconductor.org/biocLite.R")
    biocLite("missMethyl")
    library(missMethyl)
  }
  
  
  ### load data
  load(inputFilelPath)
  
  
  ### we need to compare Periodontitis from the old vs Control from the new
  mvalues_450k_virtual_sample_info <- data.frame(mvalues_450k_virtual_sample_info,
                                                 Comparison=paste(mvalues_450k_virtual_sample_info$Phenotype,
                                                                  mvalues_450k_virtual_sample_info$Platform,
                                                                  sep = "_"),
                                                 stringsAsFactors = FALSE,
                                                 check.names = FALSE)
  
  
  ### get Illumina 450k hg19 annotation data
  anno <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
  anno <- anno[match(rownames(mvalues_450k_virtual), anno$Name),
               c(1:4, 12:19, 24:(ncol(anno)-1))]
  
  
  ### get factor attributes for design matrix
  sample_type <- relevel(factor(mvalues_450k_virtual_sample_info$Comparison), ref = "Healthy_EPIC")
  ### since the data is already platform-effect-corrected, no need to consider it in differential analysis
  # sample_batch <- factor(mvalues_450k_virtual_sample_info$Platform)
  
  
  ### make a design matrix
  design <- model.matrix(~0+sample_type)
  colnames(design) <- c(levels(sample_type))
  
  
  ### fir the linear model with m value
  fit <- lmFit(mvalues_450k_virtual, design)
  
  
  ### make a contrast matrix
  contrastMat <- makeContrasts(Periodontitis_450k-Healthy_EPIC, levels = design)
  
  
  ### fit the contrasts
  fit2 <- contrasts.fit(fit, contrastMat)
  fit2 <- eBayes(fit2)
  
  
  ### get the differentially methylated CpGs
  result <- topTable(fit2, coef = 1, adjust.method="BH", number=Inf, genelist = anno)
  
  
  ### write the DMP results
  write.table(result, file = paste0(outputDir, "DMP_Periodontitis_450k_vs_Healthy_EPIC.txt"), sep = "\t", row.names = FALSE)
  
  
  ### write the DMP (FDR < fdrThreshold) results
  if(length(which(result$adj.P.Val < fdrThreshold)) > 0) {
    write.table(result[which(result$adj.P.Val < fdrThreshold),], file = paste0(outputDir, "DMP_Periodontitis_450k_vs_Healthy_EPIC_", fdrThreshold, ".txt"), sep = "\t", row.names = FALSE)
  }
  
  
  ### annotation for DMR(Differentially Methylated Region)s
  myAnno <- cpg.annotate(object = as.matrix(mvalues_450k_virtual), datatype = "array", what = "M",
                         arraytype = "450K", fdr = fdrThreshold,
                         analysis.type = "differential", design = design,
                         contrasts = TRUE, cont.matrix = contrastMat,
                         coef = "Periodontitis_450k - Healthy_EPIC")
  
  
  ### get DMRs
  if(length(which(myAnno$is.sig == TRUE)) > 0) {
    ### get DMRs
    DMRs <- dmrcate(myAnno, lambda=1000, C=2)
    
    ### extract ranges from DMRs
    results.ranges <- extractRanges(DMRs, genome = "hg19")
    
    if(length(results.ranges) > 0) {
      ### filter with Stouffer combined p-values
      results.ranges <- results.ranges[which(results.ranges$Stouffer < fdrThreshold),]
      
      ### save the DMR info
      write.table(data.frame(results.ranges), file = paste0(outputDir, "DMR_Periodontitis_450k_vs_Healthy_EPIC.txt"),
                  sep = "\t", row.names = FALSE)
    }
  }
  
  
  ### set up parameters for a plot
  pheno <- c("orange", "green", "pink", "red")
  names(pheno) <- levels(factor(mvalues_450k_virtual_sample_info$Comparison))
  cols <- pheno[as.character(factor(mvalues_450k_virtual_sample_info$Comparison))]
  
  
  ### draw DMR plots with top DMRs
  for(i in 1:dmrPrintNum) {
    ### Gingivitis - Healthy
    if(length(which(myAnno$is.sig == TRUE)) >= i) {
      ### draw DMR plots
      idx <- union(which(mvalues_450k_virtual_sample_info$Comparison == "Periodontitis_450k"),
                   which(mvalues_450k_virtual_sample_info$Comparison == "Healthy_EPIC"))
      png(paste0(outputDir, "DMR", i, "_Periodontitis_450k_vs_Healthy_EPIC.png"), width = 1800, height = 1000)
      DMR.plot(ranges=results.ranges, dmr=i, CpGs=as.matrix(mvalues_450k_virtual[,idx]), phen.col=cols[idx], what = "M",
               arraytype = "450K", pch=16, toscale=TRUE, plotmedians=TRUE, 
               genome="hg19", samps=1:length(idx))
      dev.off()
    }
  }
  
  
  ### pathway analysis with genes associated with the DMPs
  sigCpGs <- result$Name[result$adj.P.Val < fdrThreshold]
  try (
    if(length(sigCpGs) > 0) {
      ### GO
      gst_go <- gometh(sig.cpg = sigCpGs, all.cpg = result$Name, plot.bias = FALSE,
                       array.type = "450K", collection = "GO")
      gst_go <- gst_go[order(gst_go$FDR),]
      gst_go <- gst_go[which(gst_go$FDR < fdrThreshold),]
      ### KEGG
      gst_kegg <- gometh(sig.cpg = sigCpGs, all.cpg = result$Name, plot.bias = FALSE,
                         array.type = "450K", collection = "KEGG")
      gst_kegg <- gst_kegg[order(gst_kegg$FDR),]
      gst_kegg <- gst_kegg[which(gst_kegg$FDR < fdrThreshold),]
      ### write out the pathway results
      if(nrow(gst_go) > 0) {
        write.table(cbind(ID=rownames(gst_go), gst_go), file = paste0(outputDir, "GO_DMP_Periodontitis_450k_vs_Healthy_EPIC.txt"),
                    sep = "\t", row.names = FALSE)
      }
      if(nrow(gst_kegg) > 0) {
        write.table(cbind(ID=rownames(gst_kegg), gst_kegg), file = paste0(outputDir, "KEGG_DMP_Periodontitis_450k_vs_Healthy_EPIC.txt"),
                    sep = "\t", row.names = FALSE)
      }
    }
    , silent = TRUE)
  
}
