###
#   File name : FDRSignificanceQuartile2.R
#   Author    : Hyunjin Kim
#   Date      : NOV 25, 2018
#   Email     : hk2990@cumc.columbia.edu
#   Purpose   : -log10(FDR) plot with DMPs and DEGs
#
#   * THIS CODE IS DIFFERENT FROM "FDRSignificanceQuartile.R" SINCE THIS IS FOR THE OLD DATA
#     SPECIFICALLY, DIFFERENTIAL METHYLATION RESULTS FROM OLD(450K-PERIODONTITIS)+NEW(EPIC-CONTROL) GENERATED BY "Virtual450kContorls.R"
#     AND, DIFFERENTIAL EXPRESSION RESULTS FROM AFFY CHIP DATA
#
#   Instruction
#               1. Source("FDRSignificanceQuartile2.R")
#               2. Run the function "significancePlot" - specify the input files (Differential) of methylation and RNA-Seq and output directory
#               3. -log10(FDR) plot will be generated in the output directory
#
#   Example
#               > source("The_directory_of_FDRSignificanceQuartile2.R/FDRSignificanceQuartile2.R")
#               > significancePlot2(methyl450kPath="./results/DMA/DMP_Periodontitis_450k_vs_Healthy_EPIC.txt",
#                                   methyl450kRegPath="./results/DMA/DMR_Periodontitis_450k_vs_Healthy_EPIC.txt",
#                                   affyPath="./results/DEA/limma_affy_periodontitis_vs_healthy.xlsx",
#                                   geneRIFPath="./data/generifs_basic.txt",
#                                   fdrThreshold=1e-20,
#                                   importantGenePath="./results/Combined/List_-log10(FDR)_DMR_Periodontitis_Healthy_all.xlsx",
#                                   outputDir="./results/Combined/")
###

significancePlot2 <- function(methyl450kPath="//isilon.c2b2.columbia.edu/ifs/archive/shares/bisr/Papapanou/Sept_2018/DifferentialMethylation/DMP_Periodontitis_450k_vs_Healthy_EPIC.txt",
                              methyl450kRegPath="//isilon.c2b2.columbia.edu/ifs/archive/shares/bisr/Papapanou/Sept_2018/DifferentialMethylation/DMR_Periodontitis_450k_vs_Healthy_EPIC.txt",
                              affyPath="//isilon.c2b2.columbia.edu/ifs/archive/shares/bisr/Papapanou/Sept_2018/DifferentialExpression/limma_affy_periodontitis_vs_healthy.xlsx",
                              geneRIFPath="//isilon.c2b2.columbia.edu/ifs/archive/shares/bisr/Papapanou/Sept_2018/PreprocessedData/generifs_basic.txt",
                              fdrThreshold=1e-20,
                              importantGenePath="//isilon.c2b2.columbia.edu/ifs/archive/shares/bisr/Papapanou/Sept_2018/Combined/FDR_Plot/List_-log10(FDR)_DMR_Periodontitis_Healthy_all.xlsx",
                              outputDir="//isilon.c2b2.columbia.edu/ifs/archive/shares/bisr/Papapanou/Sept_2018/Combined/FDR_Plot/") {
  
  ### load libraries
  if(!require(xlsx)) {
    install.packages("xlsx")
    library(xlsx)
  }
  if(!require(ggplot2)) {
    install.packages("ggplot2")
    library(ggplot2)
  }
  if(!require(IlluminaHumanMethylation450kanno.ilmn12.hg19)) {
    source("https://bioconductor.org/biocLite.R")
    biocLite("IlluminaHumanMethylation450kanno.ilmn12.hg19")
    library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
  }
  if(!require(ggrepel)) {
    source("https://bioconductor.org/biocLite.R")
    biocLite("ggrepel")
    library(ggrepel)
  }
  
  
  ### load data
  dmp <- read.table(file = methyl450kPath, header = TRUE, sep = "\t", check.names = FALSE)
  dmr <- read.table(file = methyl450kRegPath, header = TRUE, sep = "\t", check.names = FALSE)
  de <- read.xlsx2(file = affyPath, sheetIndex = 1, stringsAsFactors = FALSE)
  
  
  ### set row names
  rownames(dmp) <- dmp$Name
  rownames(de) <- de[,1]
  de <- de[,-1]
  
  
  ### DMR
  
  ### a function to return an associated gene list vector of with a given region
  getGeneSet <- function(str) {
    ### split the string
    temp <- strsplit(as.character(str), ", ", fixed = TRUE)[[1]]
    ### remove promoter postfix
    temp <- sapply(temp, function(x) {
      return(substr(x, 1, nchar(x)-4))
    })
    
    return(temp)
  }
  
  
  ### make a data for a plot
  cor_data <- NULL
  for(i in 1:nrow(dmr)) {
    if(!is.na(dmr$overlapping.promoters[i])) {
      ### get DMR-associated gene list vector
      temp <- getGeneSet(dmr$overlapping.promoters[i])
      
      ### sometimes, there are redundant genes, so get an unique set
      temp <- unique(temp)
      
      ### a region may have more than one genes
      for(j in 1:length(temp)) {
        geneIdx <- which(de$Gene_Symbol == temp[j])
        if(length(geneIdx) > 0) {
          for(k in 1:length(geneIdx)) {
            if(!is.na(de[geneIdx[k],"adj.P.Val"])) {
              cor_data <- rbind(cor_data,
                                c(rownames(de)[geneIdx[k]],
                                  de$Gene_Symbol[geneIdx[k]],
                                  paste0("DMR", i),
                                  as.numeric(c(de[geneIdx[k],"adj.P.Val"],
                                               dmr$Stouffer[i]))))
            }
          }
        }
      }
    }
  }
  cor_data <- data.frame(cor_data)
  cor_data[,1:3] <- apply(cor_data[,1:3], 2, as.character)
  cor_data[,4:5] <- apply(cor_data[,4:5], 2, as.numeric)
  
  
  ### add CpG info
  cor_data$X6 <- NA
  for(i in 1:nrow(cor_data)) {
    ### get DMR index
    idx <- as.numeric(substring(cor_data[i,3], 4))
    
    ### narrow down with the specific chromosome
    temp <- dmp[which(as.character(dmp$chr) == as.character(dmr$seqnames[idx])), 1:4]
    
    ### get CpG names
    temp <- as.character(temp$Name[intersect(which(temp$pos >= dmr$start[idx]), which(temp$pos <= dmr$end[idx]))])
    
    ### add the info
    cor_data$X6[i] <- paste(temp, collapse = ",")
  }
  
  
  ### write out the text file
  colnames(cor_data) <- c("Probe_ID", "Gene_Name", "DMR_Name", "FDR_Expression", "FDR_Methylation", "CpG_Names")
  cor_data <- cor_data[order(cor_data[,4]),]
  write.xlsx2(cor_data, file = paste0(outputDir, "List_-log10(FDR)_", substr(basename(methylRegPath), 1, nchar(basename(methylRegPath))-4), "_all.xlsx"), row.names = FALSE)
  
  
  ### mark the significant genes
  cor_data$Label <- ""
  idx <- intersect(which(cor_data$FDR_Expression < fdrThreshold), which(cor_data$FDR_Methylation < fdrThreshold))
  cor_data$Label[idx] <- cor_data$Gene_Name[idx]
  
  
  ### make a correlation plot between DM regions and DE genes
  df <- data.frame(-log10(cor_data[,4:5]), Label=cor_data$Label, stringsAsFactors = FALSE, check.names = FALSE)
  fName <- paste0("Plot_-log10(FDR)_", substr(basename(methylRegPath), 1, nchar(basename(methylRegPath))-4), "_all.png")
  ### Because there are too many points, we have to reduce the data size - cutoff: -log10(0.00001) = 5
  df <- df[intersect(which(df$FDR_Expression > 5), which(df$FDR_Methylation > 5)),]
  ggplot(data = df, aes(x=FDR_Expression, y=FDR_Methylation)) +
    geom_point(color = "black", size = 1) +
    geom_label_repel(aes(FDR_Expression, FDR_Methylation, label = Label), color = "red", box.padding = unit(0.45, "lines")) +
    labs(title=substr(fName, 1, nchar(fName)-4)
         # ,subtitle=sprintf("S.Cor = %s, p-value = %s",
         #                  round(cor(df[,1], df[,2], use = "pairwise.complete.obs", method = "spearman"), 5),
         #                  signif(cor.test(df[,1], df[,2], method = "spearman")$p.value, 5))
         ) +
    xlab("Expression") +
    ylab("Methylation") +
    # geom_smooth(method = lm, color="gray", se=FALSE) +
    geom_vline(aes(xintercept = -log10(fdrThreshold), linetype = paste0("-log10(", fdrThreshold, ")")), color = "red") +
    geom_hline(aes(yintercept = -log10(fdrThreshold), linetype = paste0("-log10(", fdrThreshold, ")")), color = "red") +
    scale_linetype_manual(name = "FDR Threshold", values = c(2),
                          guide = guide_legend(override.aes = list(color = c("red")))) +
    theme_classic(base_size = 16)
  ggsave(filename = paste0(outputDir, fName), width = 10, height = 10)
  
  
  ### draw a plot that shows how many top genes are shared between Affy and 450k
  ### x-axis: the number of top genes
  ### y-axis: the number of SHARED top genes between Affy and 450k
  
  ### get indices based on FDR_Expression and on FDR_Mehtylation
  expTopGeneIdx <- 1:nrow(cor_data)
  methylTopGeneIdx <- order(cor_data$FDR_Methylation)
  
  ### prepare empty vectors
  shared_num <- rep(0, nrow(cor_data))
  random_num <- rep(0, nrow(cor_data))
  best_num <- rep(0, nrow(cor_data))
  
  ### fill out the data
  ### shared_num = the result
  ### random_num = when the input data are random
  ### best_num = when the input data are exactly identical
  for(i in 1:nrow(cor_data)) {
    shared_num[i] <- length(intersect(expTopGeneIdx[1:i], methylTopGeneIdx[1:i]))
    random_num[i] <- (i^2) / nrow(cor_data)
    best_num[i] <- i
  }
  
  ### save the result as PNG
  png(paste0(outputDir, "Plot_Shared_Genes_", substr(basename(methylRegPath), 1, nchar(basename(methylRegPath))-4), ".png"),
      width = 1000, height = 950, res = 130)
  plot(shared_num, main = "The number of top genes (smallest FDR) shared between Affy and 450k",
       xlab = "The number of top genes", ylab = "The number of SHARED genes")
  lines(random_num, col = "red", lty = 2)
  lines(best_num, col = "blue", lty = 2)
  legend("topleft", legend=c("Result", "Random", "Best (identical)"),
         col=c("black", "red", "blue"), lty=c(1,2,2), cex=1)
  dev.off()
  
  
  ### A table of statistics of the important genes found from the new data
  
  ### get important genes
  gList <- read.xlsx2(file = importantGenePath, sheetIndex = 1, stringsAsFactors = FALSE)
  gList[,3:4] <- apply(gList[,3:4], 2, as.numeric)
  important_genes <- gList$Gene_Name[intersect(which(gList$FDR_Expression < 0.05),
                                     which(gList$FDR_Methylation < 0.05))]
  important_genes <- unique(important_genes)
  
  ### make the table
  importantGeneTable <- NULL
  for(i in 1:length(important_genes)) {
    ### get indicies for the given gene name
    idx1 <- which(cor_data$Gene_Name == important_genes[i])
    idx2 <- which(gList$Gene_Name == important_genes[i])
    
    ### if there is no gene found in the old data, fill NA
    ### combine stastics of the given gene from both the old and the new data
    if(length(idx1) > 0) {
      for(j in 1:length(idx2)) {
        for(k in 1:length(idx1)) {
          importantGeneTable <- rbind(importantGeneTable, unlist(cbind(gList[idx2[j],], cor_data[idx1[k],-c(2,7)])))
        }
      }      
    } else {
      for(j in 1:length(idx2)) {
        importantGeneTable <- rbind(importantGeneTable, unlist(cbind(gList[idx2[j],], data.frame(t(rep("NA", 5)), stringsAsFactors = FALSE))))
      }
    }
  }
  colnames(importantGeneTable) <- c("Gene_Name", "EPIC_DMR_Name", "RNASEQ_DE_FDR", "EPIC_DMR_FDR",
                                    "EPIC_DMR_CpG_Names", "Affy_Probe_ID", "450K_DMR_Name",
                                    "Affy_DE_FDR", "450K_DMR_FDR", "450K_DMR_CpG_Names")
  importantGeneTable <- data.frame(importantGeneTable, stringsAsFactors = FALSE, check.names = FALSE)
  
  ### shared CpG sites between the old and the new
  importantGeneTable <- data.frame(importantGeneTable, Shared_CpGs=NA, Shared_CpGs_Num=NA,
                                   stringsAsFactors = FALSE, check.names = FALSE)
  for(i in 1:nrow(importantGeneTable)) {
    ### get CpG names from the EPIC and the 450K data
    epic_cpgs <- strsplit(importantGeneTable$EPIC_DMR_CpG_Names[i], split = ",", fixed = TRUE)[[1]]
    methyl450k_cpgs <- strsplit(importantGeneTable$`450K_DMR_CpG_Names`[i], split = ",", fixed = TRUE)[[1]]
    
    ### get shared CpGs between them
    shared_cpgs <- intersect(epic_cpgs, methyl450k_cpgs)
    
    ### add the info to the table
    importantGeneTable$Shared_CpGs[i] <- paste(shared_cpgs, collapse = ",")
    importantGeneTable$Shared_CpGs_Num[i] <- paste(length(shared_cpgs), "out of", length(epic_cpgs), "(EPIC) &", length(methyl450k_cpgs), "(450k)")
  }
  
  ### write out the table
  write.xlsx2(importantGeneTable, file = paste0(outputDir, "Important_Genes_Statistics_Table.xlsx"),
              row.names = FALSE)
  
  
  ### create GeneRIF text
  if(length(idx) > 0) {
    ### load geneRIF
    geneRIF <- read.table(file = geneRIFPath, header = FALSE, sep = "\t", check.names = FALSE)
    
    # *****************************************************************************
    #
    # Map Gene Symbols to Gene IDs
    #
    # geneSymbol:	A single string or a vector of strings representing gene symbol(s)
    # 
    # Returns a vector of the same size as "geneSymbol" where the i-th entry is the 
    # gene ID (as an integer) corresponsing to the i-th gene symbol in "geneSymbol". 
    # The return vector entries are named using the gene symbols. For gene symbols mapped
    # to more than one entrez ids, only the first id is returned. 
    # ATTENTION: 
    # Before used geneSymbol is first stripped of symbols which are not mapped to 
    # at least one  entrez ID. If no gene symbols remain after this stripping, a NULL 
    # value is returned.
    # *****************************************************************************
    geneSymbolToEntrezId <- function(geneSymbol){
      
      if(!require(org.Hs.eg.db)) {
        source("https://bioconductor.org/biocLite.R")
        biocLite("org.Hs.eg.db")
        library(org.Hs.eg.db)
      }
      
      keys = keys(org.Hs.eg.db,  keytype="SYMBOL")
      geneSymbol = geneSymbol[geneSymbol %in% keys]
      if (length(geneSymbol) == 0)
        return(NULL)
      
      tmp = select(org.Hs.eg.db, keys=geneSymbol, keytype="SYMBOL", columns=c("ENTREZID"))
      # Remove duplicate mappings, if any
      t = which(duplicated(tmp[,1]))
      if (length(t) > 0)
        tmp = tmp[-t, ]
      res = as.integer(tmp[,2])
      names(res) = tmp[, 1]
      return(res)
    }
    
    ### get Entrez IDs
    entIDs <- geneSymbolToEntrezId(unique(cor_data$Gene_Name[idx]))
    
    ### extract geneRIF for the specific genes
    add_info <- geneRIF[which(geneRIF$V2 %in% entIDs),c(2,5)]
    colnames(add_info) <- c("Gene_Name", "Gene_RIF")
    
    ### change Entrez IDs to gene symbols
    name_map <- names(entIDs)
    names(name_map) <- entIDs
    add_info[,1] <- name_map[as.character(add_info[,1])]
    
    ### remove duplicates
    rIdx <- duplicated.data.frame(add_info)
    add_info <- add_info[!rIdx,]
    
    ### write out the result
    write.xlsx2(add_info, file = paste0(outputDir, "GeneRIF_", substr(basename(methylRegPath), 1, nchar(basename(methylRegPath))-4), "_all.xlsx"), row.names = FALSE)
  }
  
  
  ### Since there are many genes that both differentially expressed and differentially methylated,
  ### it is possible to perform pathway analysis with those genes
  ### here, we set FDR threshold as 1e-6 for the pathway analysis
  pathwayThreshold <- 0.000001
  
  ### A function to change gene names (from human gene symbol to Entrez ID)
  convert <- function(hgnc_symbol) {
    
    ### load library
    if(!require(org.Hs.eg.db)) {
      source("https://bioconductor.org/biocLite.R")
      biocLite("org.Hs.eg.db")
      library(org.Hs.eg.db)
    }
    
    ### mapping information between Gene Symbol and Entrez ID (NCBI ID)
    map_symbol_eg <- mappedkeys(org.Hs.egSYMBOL2EG)
    list_symbol2eg <- as.list(org.Hs.egSYMBOL2EG[map_symbol_eg])
    
    ### get corresponding Entrez IDs
    entrez_id <- list_symbol2eg[hgnc_symbol]
    
    ### only keep the first element if multiple Entrez ID exists
    idx <- which(sapply(entrez_id, length) > 1)
    if(length(idx) > 0) {
      entrez_id[idx] <- sapply(idx, function(x) entrez_id[[x]][1])
    }
    entrez_id <- as.character(entrez_id)
    
    return(entrez_id)
  }
  
  # ******************************************************************************************
  # Pathway Analysis with clusterProfiler package
  # Input: geneList     = a vector of gene Entrez IDs for pathway analysis [numeric or character]
  #        org          = organism that will be used in the analysis ["human" or "mouse"]
  #                       should be either "human" or "mouse"
  #        database     = pathway analysis database (KEGG or GO) ["KEGG" or "GO"]
  #        title        = title of the pathway figure [character]
  #        pv_threshold = pathway analysis p-value threshold (not DE analysis threshold) [numeric]
  #        displayNum   = the number of pathways that will be displayed [numeric]
  #                       (If there are many significant pathways show the few top pathways)
  #        imgPrint     = print a plot of pathway analysis [TRUE/FALSE]
  #        dir          = file directory path of the output pathway figure [character]
  #
  # Output: Pathway analysis results in figure - using KEGG and GO pathways
  #         The x-axis represents the number of DE genes in the pathway
  #         The y-axis represents pathway names
  #         The color of a bar indicates adjusted p-value from the pathway analysis
  #         For Pathview Result, all colored genes are found DE genes in the pathway,
  #         and the color indicates log2(fold change) of the DE gene from DE analysis
  # ******************************************************************************************
  pathwayAnalysis_CP <- function(geneList,
                                 org,
                                 database,
                                 title="Pathway_Results",
                                 pv_threshold=0.05,
                                 displayNum=Inf,
                                 imgPrint=TRUE,
                                 dir="./") {
    
    ### load library
    if(!require(clusterProfiler)) {
      source("https://bioconductor.org/biocLite.R")
      biocLite("clusterProfiler")
      library(clusterProfiler)
    }
    if(!require(ggplot2)) {
      install.packages("ggplot2")
      library(ggplot2)
    }
    if(!require(xlsx)) {
      install.packages("xlsx")
      library(xlsx)
    }
    
    
    ### colect gene list (Entrez IDs)
    geneList <- geneList[which(!is.na(geneList))]
    
    if(!is.null(geneList)) {
      ### make an empty list
      p <- list()
      
      if(database == "KEGG") {
        ### KEGG Pathway
        kegg_enrich <- enrichKEGG(gene = geneList, organism = org, pvalueCutoff = pv_threshold)
        
        if(is.null(kegg_enrich)) {
          writeLines("KEGG Result does not exist")
        } else if(imgPrint == TRUE){
          if((displayNum == Inf) || (nrow(kegg_enrich@result) <= displayNum)) {
            result <- kegg_enrich@result
            description <- kegg_enrich@result$Description
          } else {
            result <- kegg_enrich@result[1:displayNum,]
            description <- kegg_enrich@result$Description[1:displayNum]
          }
          
          if(nrow(kegg_enrich) > 0) {
            p[[1]] <- ggplot(result, aes(x=Description, y=Count)) + labs(x="", y="Gene Counts") + 
              theme_classic(base_size = 16) + geom_bar(aes(fill = p.adjust), stat="identity") + coord_flip() +
              scale_x_discrete(limits = rev(description)) +
              guides(fill = guide_colorbar(ticks=FALSE, title="P.Val", barheight=10)) +
              ggtitle(paste0("KEGG ", title))
            
            png(paste0(dir, "kegg_", title, "_CB.png"), width = 2000, height = 1000)
            print(p[[1]])
            dev.off()
          } else {
            writeLines("KEGG Result does not exist")
          }
        }
        
        if(!is.null(kegg_enrich)) {
          return(kegg_enrich@result)  
        }
      } else if(database == "GO") {
        ### GO Pathway
        if(org == "human") {
          go_enrich <- enrichGO(gene = geneList, OrgDb = 'org.Hs.eg.db', readable = T, ont = "BP", pvalueCutoff = pv_threshold)
        } else if(org == "mouse") {
          go_enrich <- enrichGO(gene = geneList, OrgDb = 'org.Mm.eg.db', readable = T, ont = "BP", pvalueCutoff = pv_threshold)
        } else {
          go_enrich <- NULL
          writeLines(paste("Unknown org variable:", org))
        }
        
        if(is.null(go_enrich)) {
          writeLines("GO Result does not exist")
        } else if(imgPrint == TRUE) {
          if((displayNum == Inf) || (nrow(go_enrich@result) <= displayNum)) {
            result <- go_enrich@result
            description <- go_enrich@result$Description
          } else {
            result <- go_enrich@result[1:displayNum,]
            description <- go_enrich@result$Description[1:displayNum]
          }
          
          if(nrow(go_enrich) > 0) {
            p[[2]] <- ggplot(result, aes(x=Description, y=Count)) + labs(x="", y="Gene Counts") + 
              theme_classic(base_size = 16) + geom_bar(aes(fill = p.adjust), stat="identity") + coord_flip() +
              scale_x_discrete(limits = rev(description)) +
              guides(fill = guide_colorbar(ticks=FALSE, title="P.Val", barheight=10)) +
              ggtitle(paste0("GO ", title))
            
            png(paste0(dir, "go_", title, "_CB.png"), width = 2000, height = 1000)
            print(p[[2]])
            dev.off()
          } else {
            writeLines("GO Result does not exist")
          }
        }
        
        if(!is.null(go_enrich)) {
          return(go_enrich@result)  
        }
      } else {
        stop("database prameter should be \"GO\" or \"KEGG\"")
      }
    } else {
      writeLines("geneList = NULL")
    }
  }
  
  
  ### pathway analysis
  ### pathway images
  fileName <- paste0(substr(basename(methylRegPath), 1, nchar(basename(methylRegPath))-4))
  pKegg <- pathwayAnalysis_CP(geneList = convert(unique(cor_data$Gene_Name[intersect(which(cor_data$FDR_Expression < pathwayThreshold), which(cor_data$FDR_Methylation < pathwayThreshold))])),
                              org = "human", database = "KEGG",
                              title = fileName,
                              displayNum = 50, imgPrint = TRUE, dir = outputDir)
  pGo <- pathwayAnalysis_CP(geneList = convert(unique(cor_data$Gene_Name[intersect(which(cor_data$FDR_Expression < pathwayThreshold), which(cor_data$FDR_Methylation < pathwayThreshold))])),
                            org = "human", database = "GO",
                            title = fileName,
                            displayNum = 50, imgPrint = TRUE, dir = outputDir)
  ### pathway tables
  if(nrow(pKegg) > 0) {
    write.xlsx2(pKegg, file = paste0(outputDir, fileName, "_KEGG_Table.xlsx"),
                sheetName = paste0(fileName, "_KEGG"))
  }
  if(nrow(pGo) > 0) {
    write.xlsx2(pGo, file = paste0(outputDir, fileName, "_GO_Table.xlsx"),
                sheetName = paste0(fileName, "_GO"))
  }
  
}
